vignettes/communicate.Rmd

```diff
@@ -0,0 +1,349 @@
+
+```{r}
+#' @examples 
+#' add_two(1, 2)
+#' # ->
+#' sum(1, 2)
+```
+
+Then add `@keywords internal` to remove the function from the documentation index.
+If you use pkgdown, also check that it's no longer listed in `_pkgdown.yml`.
+These changes make reduce the chance of new users coming across a deprecated function, but don't prevent those who already know about it from referring to the docs.
+
+```{r}
+#' @keywords internal
+```
+
+You're now done with the docs, and it's time to add a warning when the user calls your function.
+Do this by adding call to `deprecate_warn()` on the first line of the function:
+
+```{r}
+add_two <- function(x, y) {
+  lifecycle::deprecate_warn("1.0.0", add_two(), base::sum())
+  x + y
+}
+
+add_two(1, 2)
+```
+
+`deprecate_warn()` generates user friendly messages for two common deprecation alternatives:
+
+-   Function in same package: `lifecycle::deprecate_warn("1.0.0", "fun_old()", "fun_new()")`
+
+-   Function in another package: `lifecycle::deprecate_warn("1.0.0", "old()", "package::new()")`
+
+For other cases, use the `details` argument to provide your own message to the user:
+
+```{r}
+add_two <- function(x, y) {
+  lifecycle::deprecate_warn(
+    "1.0.0", 
+    add_two(), 
+    details = "This function is a special case of sum(); use it instead."
+  )
+  x + y
+}
+
+add_two(1, 2)
+```
+
+It's good practice to test that you've correctly implemented the deprecation, testing that the deprecated function still works and that it generates a useful warning.
+Using an expectation inside an `expect_snapshot()`[^2] is a convenient way to do this:
+
+[^2]: You can learn more about snapshot testing in `vignette("snapshot", package = "testthat")`.
+
+```{r, eval = FALSE}
+test_that("add_two is deprecated", {
+  expect_snapshot({
+    x <- add_two(1, 1)
+    expect_equal(x, 2)
+  })
+})
+```
+
+If you have existing tests for the deprecated function you can suppress the warning in those tests with the `lifecycle_verbosity` option:
+
+```{r, eval = FALSE }
+test_that("add_two returns the sum of its inputs", {
+  withr::local_options(lifecycle_verbosity = "quiet")
+  expect_equal(add_two(1, 1), 2)
+})
+```
+
+And then add a separate test specifically for the deprecation.
+
+```{r, eval = FALSE}
+test_that("add_two is deprecated", {
+  expect_snapshot(add_two(1, 1))
+})
+```
+
+### Gradual deprecation
+
+For particularly important functions, you can choose to add two other stages to the deprecation process:
+
+-   `deprecated_soft()` is used before `deprecated_warn()`.
+    This function only warns (a) users who try the feature from the global environment and (b) developers who directly use the feature (when running testthat tests).
+    There is no warning when the deprecated feature is called indirectly by another package.
+    This is a gentler form of deprecation that's useful when a package changes its extension interface --- it encourages package authors to update their code without affecting the end user.
+
```

Comment:
Somehow I like this phrasing of the goal: we want to warn the person who has the power to stop using the deprecated feature. And only that person.

Reply:
Oh yeah, I like that framing!